### DISCUSSION TITLE
How to make the main window of the program full screen display, when starting

### DISCUSSION THREAD

**Discussion by moon989** [emilk/egui#477] [Q&A]
1： how to make window full screen display in native Application ; 
2:   how about in WASM ;

**Comment by parasyte** [comment]
`epi` and `egui_glium` need a new `NativeOptions` setting for this on native apps.

It is not at all possible to enter fullscreen automatically in a browser; the user _MUST_ request the app to enter full screen with some kind of interaction like a button press. See [`Element.requestFullscreen()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestFullscreen) docs for more info. This limitation is not inherent to all WASM implementations (e.g. Electron and NW.js may allow automatic fullscreen. But if you are using these types of app containers, you may as well be writing a native app).

**Reply by amandasystems** [reply]
Is this still the case?

**Reply by John-Nagle** [reply]
For native only, how do I get to full screen? Or at least borderless? Stack is

* rend3
* rend3-egui
* egui
* epi
* egui_winit_platform
* winit

Who's responsible for controlling window borders and title bars? There's a winit call available, but with all that stuff on top, I probably shouldn't be calling winit directly. I'd be happy with a full-screen window with no title bar.

(Win/Mac/Linux, not web/android, game-type application. Want a clean screen during play.)

**Comment by willser** [comment]
Following those code to into full screen immediately when window is opened
````rust
pub struct Weaver {}

impl App for Weaver {
    fn update(&mut self, ctx: &Context, frame: &mut Frame) {}
}

fn main() {
    let mut native_options = eframe::NativeOptions::default();

    native_options.maximized = true;

    eframe::run_native(
        "weaver1",
        native_options,
        Box::new(|creation_context| Box::new(Weaver::default())),
    );
}

}

````

**Reply by buxx** [reply]
Note: In current version NativeOptions struct don't have anymore a `maximized` attribute.

**Reply by YgorSouza** [reply]
These options have been moved to the viewport, so now to start the application fullscreen:

```rs
    let options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_fullscreen(true),
        ..Default::default()
    };
```

And to start maximized:

```rs
    let options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_maximized(true),
        ..Default::default()
    };
```

For more settings see [ViewportBuilder](https://docs.rs/egui/0.28.1/egui/viewport/struct.ViewportBuilder.html).

These options can also be set after the application is already running (e.g. toggle fullscreen when the user clicks a button or presses a keyboard shortcut), using [ViewportCommand](https://docs.rs/egui/0.28.1/egui/viewport/enum.ViewportCommand.html).

**Comment by dBangos** [comment]
I used wasm-bindgen to add fullscreen functionality to a button on my UI.  This is also locking the orientation,  I left that in as most people using fullscreen will probably need that too.

```rust
#[cfg(target_arch = "wasm32")]
                if ui.button("Fullscreen").clicked() {
                        let Some(window) = web_sys::window() else {
                            return;
                        };
                        let Some(document) = window.document() else {
                            return;
                        };
                    if self.wasm_fullscreen {
                        let _ = document.exit_fullscreen();

                        let Ok(screen) = window.screen() else {
                            return;
                        };
                        let _ = screen.orientation().unlock();

                        self.wasm_fullscreen = false;
                    } else {
                        let Some(element) = document.document_element() else {
                            return;
                        };
                        let _ = element.request_fullscreen();

                        let Ok(screen) = window.screen() else {
                            return;
                        };
                        let _ = screen
                            .orientation()
                            .lock(web_sys::OrientationLockType::Landscape);
                        self.wasm_fullscreen = true;
                    }
```
And added this to the cargo.toml.
```rust
web-sys = { version ="0.3.70", features =["Element", "Document", "Window","OrientationLockType", "Screen", "ScreenOrientation"]}
```
